Sources/AST/Declaration/FunctionDeclaration.swift

```diff
@@ -6,26 +6,38 @@
 //
 import Source
 import Lexer
+import ABI
 
 /// The declaration of a function.
 public struct FunctionDeclaration: ASTNode {
   public var signature: FunctionSignatureDeclaration
   public var body: [Statement]
   public var closeBraceToken: Token
+  public var isExternal: Bool
 
   public var mangledIdentifier: String?
+  public var externalSignatureHash: [UInt8]?
 
   // Contextual information for the scope defined by the function.
   public var scopeContext: ScopeContext?
 
   public init(signature: FunctionSignatureDeclaration,
               body: [Statement],
               closeBraceToken: Token,
-              scopeContext: ScopeContext? = nil) {
+              scopeContext: ScopeContext? = nil,
+              isExternal: Bool = false) {
     self.signature = signature
     self.body = body
     self.closeBraceToken = closeBraceToken
     self.scopeContext = scopeContext
+    self.isExternal = isExternal
+
+    if isExternal {
+      let args = signature.parameters.map { $0.type.rawType.name }.joined(separator: ",")
+
+      let name = signature.identifier.name
+      self.externalSignatureHash = ABI.soliditySelectorRaw(of: "\(name)(\(args))")
```

Comment:
We can probably use `mangledSignature()` as with the global router

Reply:
We decided not to use this as solidity types are not canonical types. There is still some work to do with the mapping.